grafana.tf

```diff
@@ -23,3 +30,12 @@ module "grafana" {
     }
   ]
 }
+
+resource "grafana_data_source" "thanos" {
+  name        = "thanos"
+  type        = "prometheus"
+  url         = "${cloudfoundry_route.thanos_query_internal.endpoint}:9090"
```

Comment:
I dont think this will work as there is no way that the terraform can reach the internal endpoint without being run inside CF and the network policies being configured on the fly.

I added this on in the monitoring module after i had exposed an external endpoint outside this module

Reply:
This should work fine, it's only the configuration of the data source right? Grafana, and not Terraform, will be talking to this endpoint through `proxy`

I'm hoping that the configuration of the `grafana` provider is handled correctly

Reply:
but the grafana provider needs a way to talk to grafana itself
so there is communication from where you are running the provider to the grafana instance..

Reply:
and you dont need the proxy part

I have this working here https://github.com/philips-internal/terraform-cloudfoundry-monitoring/blob/main/grafana.tf#L64

Reply:
there is no config of the grafana provider here so it would have to be configured in the caller and would need and external route to grafana to be able to talk from terraform provider to grafana.

in the module link above i have configured the provider in the module with the external route and the auto generated  credentials.

you are right that grafana will speak to prometheus internally and that will work fine. (will need a network policy in place)

Reply:
Yeah, caller needs to configure grafana providers. That's why I'm propagating the credentials setting all the way up, so they are available for this. With `proxy` I meant the data source itself. The Grafana provider only ever talks to the public Grafana endpoint. However, the data source internal to Grafana (i.e. `thanos-query`) is accessed internally by Grafana.

Reply:
How does the dependency tree like this setup?

Feels a little off that we expect the caller to setup the provider dynamically based on exports of the module so that they can then be used inside the module..

We have to assume the grafana needs a public endpoint anywhere here so we should just include the provisioning of that inside the module.

Reply:
I'm reading this now: https://www.terraform.io/docs/language/modules/develop/providers.html

What I understand is that it is allowed to internalize the provider config. You are right, it's kind of weird to have the caller be responsible for providing the `grafana` config. I'll move the code inside the `thanos` module to test.. Note that `Grafana` is now no longer an optional thing, this removes al lot of conditionals.. I'm on the fence on the backing database..

Reply:
Works as expected